File: app/rotas/configuracao.py
"""/config - parametros operacionais, backup e exportacao."""

from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import FileResponse
from sqlalchemy import select

from app.config import VERSAO, obter_config
from app.dependencias import SessaoDep, UsuarioDep
from app.modelos import AutoridadeDestinataria, SetorEmissor
from app.servicos.rbac import PermissaoNegada
from app.servicos import auditoria, backup as servico_backup, sei
from app.servicos.processo import SLA_PADRAO, gravar_sla, sla_vigente
from app.web import pagina

rotas = APIRouter(tags=["config"])

# Quem abre `/config`, e por quê são duas permissões e não uma.
#
# A tela guarda coisas de duas naturezas. Uma é **parâmetro do processo**: o SLA
# que pinta o vermelho de toda lista, o setor emissor e a autoridade
# destinatária, que são congelados no parecer na emissão (RN-15). A outra é
# **operação da máquina**: onde o banco e os anexos moram, o backup cifrado e a
# exportação total — a continuidade do negócio.
#
# Exigir só `processo.ver` deixava a segunda trancada para o único perfil que
# existe para ela: `admin_ti` tem `backup.executar` e não tem `processo.ver`, e
# a base normativa dele diz por quê ("sem acesso ao conteúdo técnico: só conta,
# senha, auditoria e backup"). Quem faz backup não alcançava a tela do backup, e
# a decisão estava fixada em teste desde a 1.0.
#
# A saída NÃO é abrir a tela inteira. Abrir demais é defeito do mesmo tamanho:
# entregar o SLA e os nomes e cargos da autoridade signatária ao administrador de
# TI contraria a promessa que torna seguro dar a ele o banco. Então a ROTA aceita
# qualquer uma das duas — quem tem uma das duas tem o que fazer aqui — e a TELA
# reparte: os cartões de processo pedem `processo.ver`, os de máquina abrem para
# quem entrou. É a mesma repartição que a tela já fazia dentro do cartão de
# backup, onde o botão sempre pediu `backup.executar`.
PERMISSOES_CONFIG: tuple[str, ...] = ("processo.ver", "backup.executar")


def _exigir_config(usuario) -> None:
    """Qualquer uma de `PERMISSOES_CONFIG` — nunca as duas.

    `dependencias.exigir` é E, não OU: ela existe para somar exigências. O padrão
    de "qualquer uma destas" já vive em `modulos.Item.permissao`, e é o mesmo que
    `/usuarios` e `/pendencias` usam. A porta da tela (a engrenagem do bloco do
    usuário, em `base.html`) repete esta tupla — se um dia ela divergir, o menu
    volta a oferecer porta trancada.
    """
    if not any(usuario.pode(codigo) for codigo in PERMISSOES_CONFIG):
        raise PermissaoNegada(
            PERMISSOES_CONFIG[0],
            "A configuração do sistema abre para quem instrui processo "
            "(`processo.ver`) ou para quem opera a máquina (`backup.executar`).",
        )


@rotas.get("/config")
def tela(
    request: Request,
    s: SessaoDep,
    usuario: UsuarioDep,
    mensagem: str | None = None,
    erro: str | None = None,
):
    _exigir_config(usuario)
    cfg = obter_config()
    # o que é do processo só é CONSULTADO por quem pode vê-lo: consultar para
    # depois esconder no template é o convite a alguém remover o `{% if %}` e não
    # perceber que o dado já estava na mão
    do_processo = usuario.pode("processo.ver")
    return pagina(
        request,
        "paginas/config.html",
        usuario=usuario,
        cfg=cfg,
        versao=VERSAO,
        do_processo=do_processo,
        sla=sla_vigente(s) if do_processo else {},
        sla_padrao=SLA_PADRAO,
        setores=list(s.execute(select(SetorEmissor)).scalars()) if do_processo else [],
        destinatarios=(
            list(s.execute(select(AutoridadeDestinataria)).scalars())
            if do_processo
            else []
        ),
        passos_sei=sei.PASSOS_INCLUSAO if do_processo else (),
        versao_sei=sei.VERSAO_SEI_ALVO,
        mensagem=mensagem,
        erro=erro,
    )


@rotas.post("/config/sla")
async def salvar_sla(request: Request, s: SessaoDep, usuario: UsuarioDep):
    """RN-16 — o prazo de cada coluna é decisão do setor, não constante de código."""
    dados = await request.form()
    valores = {
        coluna: int(dados[f"sla_{coluna}"])
        for coluna in SLA_PADRAO
        if str(dados.get(f"sla_{coluna}", "")).strip().isdigit()
    }
    try:
        gravar_sla(s, valores, usuario)
    except PermissaoNegada as falha:
        return tela(request, s, usuario, erro=str(falha))
    s.commit()
    return tela(request, s, usuario, mensagem="SLA atualizado.")


@rotas.post("/config/backup")
def executar_backup(request: Request, s: SessaoDep, usuario: UsuarioDep):
    usuario.exigir("backup.executar")
    try:
        resultado = servico_backup.fazer_backup()
    except servico_backup.DestinoProibido as falha:
        return tela(request, s, usuario, erro=str(falha))
    auditoria.registrar(
        s,
        entidade="sistema",
        entidade_id=0,
        tipo_evento="BACKUP_EXECUTADO",
        descricao=(
            f"{resultado.arquivo.name} ({resultado.tamanho} bytes, cifrado, "
            f"banco + {resultado.arquivos} arquivo(s) de anexos/documentos)."
        ),
        usuario=usuario,
    )
    s.commit()
    return tela(
        request,
        s,
        usuario,
        mensagem=(
            f"Backup cifrado gerado: {resultado.arquivo.name} — banco e "
            f"{resultado.arquivos} arquivo(s) de anexos e documentos."
        ),
    )


# POST, e nao GET. O que sai daqui e um zip EM CLARO com o banco inteiro, os
# CSVs nominais (nome, SIAPE, cargo, unidade, agente nocivo, fundamentacao) e
# todos os anexos, inclusive os comprovantes de EPI com a assinatura manuscrita
# do titular — e ate a 1.31.0 bastava um GET para produzi-lo.
#
# `SameSite=lax` NAO cobre isto. Ele existe para nao mandar o cookie em POST de
# outra origem, e manda em navegacao de topo — que e exatamente o que um
# `<img src>` nao faz mas um link, um `<meta refresh>` ou um `window.open` de
# uma pagina qualquer fazem. Bastava alguem com a permissao `exportar` abrir
# essa pagina, logado, para o pacote ser gerado e gravado em
# `CSSO_BACKUP_DESTINO` — que a POLITICA_RETENCAO ja preve que seja uma pasta de
# rede. Numa maquina so isso era teorico; com a equipe na rede, nao e.
#
# Virar POST devolve o `SameSite=lax` ao seu trabalho: o cookie deixa de ser
# enviado, e a rota volta a exigir que quem clicou estivesse nesta tela.
@rotas.post("/config/exportar-tudo")
def exportar(s: SessaoDep, usuario: UsuarioDep):
    usuario.exigir("exportar")
    alvo = servico_backup.exportar_tudo(s)
    auditoria.registrar(
        s,
        entidade="sistema",
        entidade_id=0,
        tipo_evento="EXPORTACAO_COMPLETA",
        descricao=alvo.name,
        usuario=usuario,
    )
    s.commit()
    return FileResponse(alvo, filename=alvo.name, media_type="application/zip")

File: nuvem/src/rotas/configuracao.ts
/**
 * /config — parâmetros operacionais, backup e exportação.
 * Porte de `app/rotas/configuracao.py`.
 *
 * Quem abre `/config`, e por que são duas permissões e não uma: a tela guarda
 * **parâmetro do processo** (SLA, setor emissor, autoridade destinatária) e
 * **operação da máquina** (onde o dado mora, backup, exportação). A ROTA aceita
 * qualquer uma das duas; a TELA reparte — os cartões de processo pedem
 * `processo.ver`, os de máquina abrem para quem entrou.
 */
import { Hono } from "hono";
import type { Ambiente, Ctx } from "../nucleo/contexto.js";
import { formulario, usuarioLogado } from "../dependencias.js";
import { obterConfig, VERSAO } from "../config.js";
import { autoridade_destinataria, setor_emissor } from "../db/esquema/index.js";
import { PermissaoNegada, type UsuarioAtual } from "../servicos/rbac.js";
import * as auditoria from "../servicos/auditoria.js";
import * as servico_backup from "../servicos/backup.js";
import * as sei from "../servicos/sei.js";
import { SLA_PADRAO, gravar_sla, sla_vigente } from "../servicos/processo.js";
import { obterArmazenamento, pastaLocal } from "../servicos/armazenamento.js";
import { dicionario } from "../dicionario.js";
import "../servicos/pendencias.js"; // liga o sino da casca
import { marcarDownload, pagina, redirecionar } from "../web.js";

export const rotas = new Hono<Ambiente>();

export const PERMISSOES_CONFIG: readonly string[] = ["processo.ver", "backup.executar"];

/** Qualquer uma de `PERMISSOES_CONFIG` — nunca as duas. */
function _exigir_config(usuario: UsuarioAtual): void {
  if (!PERMISSOES_CONFIG.some((codigo) => usuario.pode(codigo))) {
    throw new PermissaoNegada(
      PERMISSOES_CONFIG[0]!,
      "A configuração do sistema abre para quem instrui processo " +
        "(`processo.ver`) ou para quem opera a máquina (`backup.executar`).",
    );
  }
}

/**
 * Onde o dado mora NESTA instalação. No Python eram caminhos de disco
 * (`cfg.caminho_banco`, `dir_anexos`...); na nuvem são o Postgres (sem a senha
 * da URL), o bucket do Storage e os prefixos de backup e exportação.
 */
function _onde_mora() {
  const cfg = obterConfig();
  let banco = "—";
  try {
    const u = new URL(cfg.bancoUrl);
    banco = `PostgreSQL ${u.hostname}:${u.port || "5432"}${u.pathname}`;
  } catch {
    /* URL ausente: o aviso de configuração já diz */
  }
  const arquivos = cfg.supabaseUrl
    ? `Supabase Storage, bucket privado “${cfg.bucketArquivos}”`
    : `pasta local ${pastaLocal()}`;
  return {
    banco,
    anexos: `${arquivos} — anexos/`,
    documentos: `${arquivos} — documentos/`,
    backup: `${arquivos} — ${servico_backup.PREFIXO_BACKUP}/ (cifrado)`,
    exportacao: `${arquivos} — ${servico_backup.PREFIXO_EXPORTACAO}/ (em claro)`,
  };
}

async function _tela(c: Ctx, usuario: UsuarioAtual, recado: { mensagem?: string | null; erro?: string | null } = {}) {
  _exigir_config(usuario);
  const tx = c.get("tx");
  const cfg = obterConfig();
  // o que é do processo só é CONSULTADO por quem pode vê-lo
  const do_processo = usuario.pode("processo.ver");
  return pagina(c, "paginas/config.html", usuario, {
    cfg: { ambiente: cfg.ambiente, epi_local_retirada: cfg.epiLocalRetirada },
    onde: _onde_mora(),
    versao: VERSAO,
    do_processo,
    sla: dicionario(do_processo ? await sla_vigente(tx) : {}),
    sla_padrao: SLA_PADRAO,
    setores: do_processo ? await tx.select().from(setor_emissor) : [],
    destinatarios: do_processo ? await tx.select().from(autoridade_destinataria) : [],
    passos_sei: do_processo ? sei.PASSOS_INCLUSAO : [],
    versao_sei: sei.VERSAO_SEI_ALVO,
    mensagem: recado.mensagem ?? null,
    erro: recado.erro ?? null,
  });
}

rotas.get("/config", async (c) => {
  const usuario = await usuarioLogado(c);
  return _tela(c, usuario, { mensagem: c.req.query("mensagem") ?? null, erro: c.req.query("erro") ?? null });
});

/** RN-16 — o prazo de cada coluna é decisão do setor, não constante de código. */
rotas.post("/config/sla", async (c) => {
  const usuario = await usuarioLogado(c);
  const f = await formulario(c);
  const valores: Record<string, number> = {};
  for (const coluna of Object.keys(SLA_PADRAO)) {
    const bruto = f.texto(`sla_${coluna}`).trim();
    if (/^\d+$/.test(bruto)) valores[coluna] = Number(bruto);
  }
  try {
    await gravar_sla(c.get("tx"), valores, usuario);
  } catch (falha) {
    if (!(falha instanceof PermissaoNegada)) throw falha;
    return _tela(c, usuario, { erro: falha.message });
  }
  return _tela(c, usuario, { mensagem: "SLA atualizado." });
});

rotas.post("/config/backup", async (c) => {
  const usuario = await usuarioLogado(c);
  usuario.exigir("backup.executar");
  const tx = c.get("tx");
  let resultado: servico_backup.ResultadoBackup;
  try {
    resultado = await servico_backup.fazer_backup(tx);
  } catch (falha) {
    if (!(falha instanceof servico_backup.DestinoProibido)) throw falha;
    return _tela(c, usuario, { erro: falha.message });
  }
  await auditoria.registrar(tx, {
    entidade: "sistema",
    entidade_id: 0,
    tipo_evento: "BACKUP_EXECUTADO",
    descricao:
      `${resultado.nome} (${resultado.tamanho} bytes, cifrado, ` +
      `banco (${resultado.tabelas} tabelas) + ${resultado.arquivos} arquivo(s) de anexos).`,
    usuario,
  });
  return _tela(c, usuario, {
    mensagem: `Backup cifrado gerado: ${resultado.chave} — banco e ${resultado.arquivos} arquivo(s) de anexos.`,
  });
});

/**
 * POST, e não GET: o que sai daqui é um zip EM CLARO com o banco inteiro.
 *
 * Na nuvem a resposta passa fácil dos 6 MB de corpo da função do Netlify: com
 * o Storage, a cópia gravada é entregue por URL assinada de 60 s; na pasta
 * local, o zip sai direto na resposta.
 */
rotas.post("/config/exportar-tudo", async (c) => {
  const usuario = await usuarioLogado(c);
  usuario.exigir("exportar");
  const tx = c.get("tx");
  const alvo = await servico_backup.exportar_tudo(tx);
  await auditoria.registrar(tx, {
    entidade: "sistema",
    entidade_id: 0,
    tipo_evento: "EXPORTACAO_COMPLETA",
    descricao: alvo.nome,
    usuario,
  });
  const arm = obterArmazenamento();
  if (arm.urlAssinada) return redirecionar(c, await arm.urlAssinada(alvo.chave, 60, alvo.nome));
  marcarDownload(c);
  return c.body(alvo.bytes as unknown as ArrayBuffer, 200, {
    "content-type": "application/zip",
    "content-disposition": `attachment; filename="${alvo.nome}"`,
  });
});
